core/src/main/scala/io/projectglow/gff/GffDataSource.scala

```diff
@@ -0,0 +1,245 @@
+              col(f.name).cast(f.dataType)
+            ).otherwise(lit(null))
+          )
+        case f => // TODO: To be completed (also rectify column names)
+          df.withColumn(
+            f.name,
+            element_at(col(attributesMapColumnName), f.name)
+          )
+      }
+    )
+      .drop(attributesField.name, attributesMapColumnName) // TODO: Complete the option of keeping attributes field
+      .rdd
+  }
+}
+
+object GffDataSource {
+
+  def inferSchema(sqlContext: SQLContext, path: String): StructType = {
+    val spark = sqlContext.sparkSession
+    val csvDf = spark
+      .read
+      .option("sep", COLUMN_DELIMITER)
+      .schema(GffDataSource.gffCsvSchema)
+      .csv(path)
+
+
+    val attributeTags = convertAttributesToMap(
+      filterCommentsAndFasta(
+        csvDf
+      )
+    )
+      .select(attributesMapColumnName)
+      .withColumn(
+        attributesMapColumnName,
+        explode(
+          map_keys(
+            col(attributesMapColumnName)
+          )
+        )
+      )
+      .agg(
+        collect_set(
+          attributesMapColumnName
+        )
+      )
+      .collect()(0).getAs[Seq[String]](0)
+
+    // generate the schema. The field names will be case and underscore insensitive
+    val officialAttributeFields =
+      gffOfficialAttributeFields.foldLeft(Seq[StructField]()) { (s, f) =>
+        if (attributeTags
+          .map(t => deUnderscore(t.toLowerCase))
+          .contains(f.name.toLowerCase)) {
+          s :+ f
+        } else {
+          s
+        }
+      }
+
+    val remainingTags = attributeTags
+      .filter(
+        t =>
+          !officialAttributeFields
+            .map(_.name.toLowerCase)
+            .contains(deUnderscore(t.toLowerCase)) && !t.isEmpty
+      )
+
+    val unofficialAttributeFields = remainingTags.foldLeft(Seq[StructField]()) { (s, t) =>
+      s :+ StructField(t, StringType)
+    }
+
+    StructType(
+      gffBaseSchema.fields ++ officialAttributeFields ++ unofficialAttributeFields
+    )
+
+  }
+
+  def filterCommentsAndFasta(df: DataFrame): DataFrame = {
+    df.where(
+        !gffCsvSchema.fieldNames.drop(1).foldLeft(lit(true))(
+          (f1, f2) =>
+            f1 && isnull(col(f2))
+        ) &&
+          substring(col(seqIdField.name), 1, 1) =!= lit(COMMENT_IDENTIFIER) &&
+          substring(col(seqIdField.name), 1, 1) =!= lit(CONTIG_IDENTIFIER)
+      )
+  }
+
+  def convertAttributesToMap(df: DataFrame): DataFrame = {
+    df.withColumn(
+        attributesMapColumnName,
+        expr(
+          s"""str_to_map(${attributesField.name},
+             | "$ATTRIBUTES_DELIMITER",
+             |  "$GFF3_TAG_VALUE_DELIMITER")""".stripMargin
+        )
+      )
+  }
+
+
+
+  def deUnderscore(s: String): String = {
+    s.replaceAll("_", "")
+  }
+
+  private[gff] val gffCsvSchema = StructType(
+    gffBaseSchema
+      .fields
+      .toSeq
+      .map(f => StructField(f.name, StringType))
```

Comment:
Why do you have to parse everything as a string at first? You can't parse them as the proper data types?

Reply:
I used nullValue option of csv reader and was able to do it.